### PR TITLE
Use task based learning instead

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
Since CUDA.jl can handle several GPUs from a single process via tasks and since according to @maleadt it will be easier to write for, let's move away from using processes to train these models. This PR adds the first building blocks for the task based mechanism, and it can potentially be very useful. One of the big ones would be if we can avoid all the serialisation cost we pay currently to sync our gradients. If that can be done via a NCCL backend, it should save at least some cost in terms of the data movement. it should also technically make it easier to implement gradient overlapping to speed up training.

Currently it seems that the `prepare_training` function only allocates the models on the same GPU, whereas the data is distributed as expected.

### CODE REVIEW

File: src/ddp_tasks.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  data = minibatch(nothing, key, nsamples = nsamples, class_idx = 1:1000)
+  shards = Flux.Data.DataLoader(data, batchsize = nsamples ÷ length(devices))
+  ds = Dict()
+  @sync for (dev, shard) in zip(devices, shards)
+    Threads.@spawn begin
+      ds[dev] = CUDA.device!(dev) do
+        gpu(resnet),
+        Flux.Data.DataLoader(gpu, shard, batchsize = 48)
+      end
+    end
```
Comment (DhairyaLGandhi) [inline src/ddp_tasks.jl:40]: This is the code in question

**Comment by DhairyaLGandhi** [comment]
Sitrep: updating to CUDA 3.8 (from 3.7) helped evaluate gradients on every GPU :tada:

But due to task scheduling, IO (or CPU related tasks) are pausing the GPU tasks (which we don't want happening). This creates hangs in the GPU tasks, slowing it down to very unusable levels. We need to guard those tasks from getting paused somehow. I have tried adding `yield` statements in the `minibatch` function and `lock`s around the task doing the GPU compute but I might be missing something. cc @chriselrod

**Comment by DhairyaLGandhi** [comment]
Another update: overscheduling and manually starting Julia with more threads than `-tauto` helps with tasks blocking on cpu activity a little bit, but doesn't actually solve it. We need every GPU to sync with every gradient step and that still causes slowdowns. Tasks still grab on to new work eagerly and don't let other GPUs do their thing.

I'm also able to copy over the gradients with a latency of ~1ms on trantor, but `fmapstructure` doesn't preserve the structure and collapses some fields (see `destruct`). Ideally it should return the exact same tree as `gradient`.

**Comment by DhairyaLGandhi** [comment]
Interestingly, using the `MARKER` method result in an illegal memory access. Not sure what about it is causing it because we shouldn't be doing anything to do with the GPU, save using the devices as the keys.
